bpf/lib/fib.h

```diff
@@ -323,24 +259,32 @@ fib_lookup_v4(struct __ctx_buff *ctx, struct bpf_fib_lookup_padded *fib_params,
 
 static __always_inline int
 fib_redirect_v4(struct __ctx_buff *ctx, int l3_off,
-		struct iphdr *ip4, const bool needs_l2_check,
-		__s8 *fib_err, int iif, int *oif)
+		struct iphdr *ip4 __maybe_unused, const bool needs_l2_check,
+		__s8 *fib_err __maybe_unused, int *oif)
 {
-	struct bpf_fib_lookup_padded fib_params = {
-		.l = {
-			.family		= AF_INET,
-			.ifindex	= iif,
-			.ipv4_src	= ip4->saddr,
-			.ipv4_dst	= ip4->daddr,
-		},
-	};
+	struct bpf_fib_lookup_padded fib_params __maybe_unused = {0};
+	void *data_end;
+	void *data;
+	__s8 skip_fib __maybe_unused = BPF_FIB_LKUP_RET_NO_NEIGH;
 	int ret;
 
 	ret = ipv4_l3(ctx, l3_off, NULL, NULL, ip4);
 	if (unlikely(ret != CTX_ACT_OK))
 		return ret;
 
-	return fib_redirect(ctx, needs_l2_check, &fib_params, fib_err, oif);
+	/* we touched ttl above, revalidate for any header reads/writes. */
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
+
```

Comment:
Can we reshuffle the code so that this revalidation isn't needed? Do the FIB lookup first, then adjust the TTL, then do the redirect?

Otherwise this should respect the passed-in `l3_off`, and use `revalidate_data_l3_off()`.

Reply:
Stared at the code for a bit and I think I see what you're saying. 

My initial concern was, we touch TTL regardless, and we'll need to revalidate if we update mac addresses in `fib_do_redirect`. But after inspection of the helper methods we call in `fib_do_redirect` such as `eth_store_daddr`, it appears they do their own revalidation internally (getting the latest data and data_end pointers). 

Are we on the same page here about why the reshuffling works?

Reply:
yep, a sequence like

```
fib_lookup_v4()
ipv4_l3()
fib_do_redirect()
```

should work, as the `fib_lookup()` has no reason to invalidate the context pointers. It operates on the FIB params block we provide.

Reply:
implemented